server/datastore/mysql/migrations/tables/20240813103800_SupportMultipleABMTokens.go

```diff
@@ -0,0 +1,192 @@
+	-- (NULL means "no team").
+	macos_default_team_id  int(10) UNSIGNED DEFAULT NULL,
+	ios_default_team_id    int(10) UNSIGNED DEFAULT NULL,
+	ipados_default_team_id int(10) UNSIGNED DEFAULT NULL,
+
+	created_at             TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
+	updated_at             TIMESTAMP NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
+
+	PRIMARY KEY (id),
+
+	UNIQUE KEY idx_abm_tokens_organization_name (organization_name),
+
+	CONSTRAINT fk_abm_tokens_macos_default_team_id
+		FOREIGN KEY (macos_default_team_id) REFERENCES teams (id) ON DELETE SET NULL,
+	CONSTRAINT fk_abm_tokens_ios_default_team_id
+		FOREIGN KEY (ios_default_team_id) REFERENCES teams (id) ON DELETE SET NULL,
+	CONSTRAINT fk_abm_tokens_ipados_default_team_id
+		FOREIGN KEY (ipados_default_team_id) REFERENCES teams (id) ON DELETE SET NULL
+)`)
+	if err != nil {
+		return fmt.Errorf("failed to create table abm_tokens: %w", err)
+	}
+
+	_, err = tx.Exec(`
+ALTER TABLE host_dep_assignments
+	ADD COLUMN abm_token_id int(10) UNSIGNED NULL,
+	ADD CONSTRAINT fk_host_dep_assignments_abm_token_id
+		FOREIGN KEY (abm_token_id) REFERENCES abm_tokens(id) ON DELETE SET NULL
+`)
+	if err != nil {
+		return fmt.Errorf("failed to alter table host_dep_assignments: %w", err)
+	}
+
+	// migrate the existing ABM token (if any) to the new abm_tokens table
+	const getABM = `
+SELECT
+	value
+FROM
+	mdm_config_assets
+WHERE
+	name = ? AND deletion_uuid = ''
+`
+	txx := sqlx.Tx{Tx: tx, Mapper: reflectx.NewMapperFunc("db", sqlx.NameMapper)}
+	var token []byte
+	if err := txx.Get(&token, getABM, fleet.MDMAssetABMTokenDeprecated); err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			// nothing to migrate, exit early
+			return nil
+		}
+		return fmt.Errorf("selecting team configs: %w", err)
+	}
+
+	// get the current ABM configuration from the app config
+	const getABMCfg = `
+SELECT
+	json_value->"$.mdm"
+FROM
+	app_config_json
+LIMIT 1
+`
+	var raw json.RawMessage
+	if err := txx.Get(&raw, getABMCfg); err != nil {
+		return fmt.Errorf("select MDM config from app_config_json: %w", err)
+	}
+
+	var (
+		abmTermsExpired bool
+		abmDefaultTeam  string
+	)
+	// decode if we did get an object
+	if len(raw) > 0 && raw[0] == '{' {
+		var config map[string]interface{}
+		if err := json.Unmarshal(raw, &config); err != nil {
+			return fmt.Errorf("unmarshal appconfig: %w", err)
+		}
+
+		if s, ok := config["apple_bm_default_team"].(string); ok {
+			abmDefaultTeam = s
+		}
+		if b, ok := config["apple_bm_terms_expired"].(bool); ok {
+			abmTermsExpired = b
+		}
+	}
+
+	var defaultTeamID *uint
+	if abmDefaultTeam != "" {
+		// get the default team id
+		var id uint
+		if err := txx.Get(&id, "SELECT id FROM teams WHERE name = ?", abmDefaultTeam); err != nil {
+			if !errors.Is(err, sql.ErrNoRows) {
+				return fmt.Errorf("select default ABM team id: %w", err)
+			}
+		}
+		// only use it if the team exists
+		if id > 0 {
+			defaultTeamID = &id
```

Comment:
We default to "no team" if the configured default team name in the AppConfig does not exist (I believe this is the fallback behaviour anyway when the cron job runs to enroll hosts and the default team does not exist).

It is:

https://github.com/fleetdm/fleet/blob/f78761dbf6480de3a89c1765d027ee3770e4fb78/server/datastore/mysql/apple_mdm.go#L914-L922